dart/gui/osg/WorldNode.hpp

```diff
@@ -67,7 +68,8 @@ class WorldNode : public ::osg::Group
   friend class Viewer;
 
   /// Default constructor
-  explicit WorldNode(std::shared_ptr<dart::simulation::World> _world = nullptr);
+  /// Shadows are disabled by default
+  explicit WorldNode(std::shared_ptr<dart::simulation::World> _world = nullptr, ::osg::ref_ptr<osgShadow::ShadowTechnique> shadowTechnique = nullptr);
```

Comment:
I'm still a little bit leery about having `shadowTechnique = nullptr` secretly mean "Let us choose a default shading technique for you".

However, we do already have a precedent of this pattern where passing in a `nullptr` for `ResourceRetriever` means that we will use a default `LocalResourceRetriever`.

I might feel better about this pattern if we rename the argument from `shadowTechnique` to something like `alternativeTechnique`. That will at least tell the user that they are specifying an alternative rather than specifying the technique to use. Then passing a `nullptr` naturally says "I don't want to use an alternative", and so we apply the default shading technique.

Sorry if this sounds petty, but without *something* along these lines, a user would have to look into the source code to actually understand how these functions should be used.

Reply:
> Sorry if this sounds petty, but without something along these lines, a user would have to look into the source code to actually understand how these functions should be used.

No worries. I understand what you say. Nevertheless, in this version of the code, `nullptr` means no shadow (not a default shadow behavior).. The user has to specify their own `ShadowTechnique`. I could create a static default technique if you like..

Reply:
Oh, sorry, I didn't notice that part of the recent change.

I actually liked providing a default shadow technique; I just wasn't comfortable with the semantics of "giving a `nullptr` technique means we should use a default". Sorry for not communicating that well.

I'm not clear on how a static default would work, since it seems to need information about the Viewer that's using it. I guess we could make it a factory function that takes a `Viewer*` argument and returns a `ref_ptr<ShadowTechnique>`? We wouldn't be able to use that function as the default argument for this constructor, but it could at least make it easy for users to enable shadows. We could add a comment in the documentation of this function that refers to the default ShadowTechnique factory.

Since the shadows conflict with the ImGui feature, I suppose we should have shadows turned off by default anyway. It would've been nice to have shadows on by default, but we can tackle that later once we've fixed the ImGui conflict.

Reply:
> I'm not clear on how a static default would work, since it seems to need information about the Viewer that's using it. I guess we could make it a factory function that takes a Viewer* argument and returns a ref_ptr<ShadowTechnique>? We wouldn't be able to use that function as the default argument for this constructor, but it could at least make it easy for users to enable shadows. We could add a comment in the documentation of this function that refers to the default ShadowTechnique factory.

That's exactly what I had in mind.. A static member function that creates a default `ShadowTechnique` (given a `Viewer*`) that users can use.. I'll do that asap..

> ince the shadows conflict with the ImGui feature, I suppose we should have shadows turned off by default anyway. It would've been nice to have shadows on by default, but we can tackle that later once we've fixed the ImGui conflict.

I agree on this and that's the reason I left the shadows disabled by default.. :smile:

Reply:
I was also thinking of modifying the `osgOperationalSpaceControl` example to be able to enable/disable shadows interactively (with a simple event handler). What do you think?

Reply:
That sounds like a great place to put an example of this, thanks!